### PR TITLE
Add What’s in the Task? documentation article

### PR DESCRIPTION
**Pull request by mbasmanova**

### CODE REVIEW

File: velox/docs/develop/task.rst [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+===================
+
+**TL;DR** Some operators need to coordinate their work with other operators. Table
+scan operators running in parallel need to pull splits from a shared queue. The
+HashProbe operator needs to wait for the HashBuild operator to create a hash
+table. Multiple HashBuild operators building different sections of the hash
+table in parallel need to coordinate to combine the pieces into one final hash
+table. Exchange operators running in parallel need to pull data from shared
+remote sources. Local exchange operators need to write and read data into
+shared data partitions. Task is maintaining shared state and enabling
```
Comment (majetideepak) [inline velox/docs/develop/task.rst:12]: Adding the following makes it sounds better to me `Task coordinates these by maintaining a shared state .....`

File: velox/docs/develop/task.rst [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+into one or more Operators. Some plan nodes are converted into multiple
+Operators of different types. For example, TableScanNode is converted into
+multiple TableScan operators, one per thread of execution (Driver), while
+HashJoinNode is converted into multiple HashProbe operators and multiple
+HashBuild operators. The number of HashProbe and HashBuild operators may or may
+not match.
+
+Task creates Drivers and Operators, then transfers the ownership of Operators to
+the corresponding Drivers. Each Driver is assigned a pipeline ID and a driver
+ID. Pipeline IDs are zero-based sequential numbers. Driver IDs are also
```
Comment (majetideepak) [inline velox/docs/develop/task.rst:79]: Should we describe `Operator ID` also?

Comment (majetideepak) [inline velox/docs/develop/task.rst:79]: Should we mention Operator ID also?

File: velox/docs/develop/task.rst [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+operator always runs single-threaded. Consumer pipelines produce sorted output
+and typically run single-threaded as well.
+
+.. image:: images/task-local-merge-plan.png
+    :width: 400
+    :align: center
+
+To summarize, a query plan is converted into multiple pipelines and each
+pipeline is executed using multiple Drivers. A given plan node is converted
+into one or more Operators. Some plan nodes are converted into multiple
```
Comment (majetideepak) [inline velox/docs/develop/task.rst:70]: `Operators` should be the small case `operators`? There are a couple of others like this as well.

Comment (majetideepak) [inline velox/docs/develop/task.rst:70]: `Operators` should be the small case `operators`? There are a couple of others like this as well.

File: velox/docs/develop/task.rst [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+within a pipeline, but repeat across pipelines. Each Operator is provided with
+the plan node ID of the corresponding plan node and a DriverCtx specifying
+pipeline ID and driver ID. Operators of the same type with matching pipeline
+IDs are called peers. These operators are running concurrently in different
+Drivers.
+
+Splits
+------
+
+Task is responsible for receiving splits from the application, storing these and
```
Comment (majetideepak) [inline velox/docs/develop/task.rst:90]: `storing these, and`. A comma is missing.

Comment (majetideepak) [inline velox/docs/develop/task.rst:90]: `storing these, and`. A comma is missing.

File: velox/docs/develop/task.rst [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+providing access to the operators.
+
+Task::addSplit(planNodeId, split) API allows the application to add a split for
+a particular plan node. The application calls addSplit() API multiple times to
+add multiple splits for multiple plan nodes. Task::noMoreSplits() API allows
+the application to signal that it added all the splits. The Task will complete
+once it processed all the splits and received no-more-splits signal.
+
+Task stores the splits in queues: one queue per plan node ID.
```
Comment (majetideepak) [inline velox/docs/develop/task.rst:99]: `planNodeId` ?

Comment (majetideepak) [inline velox/docs/develop/task.rst:99]: `planNodeId` ?

**Review by majetideepak — APPROVED** [review]
@mbasmanova, very nice explanation! Thanks for doing this. I left some minor nits.

**Comment by mbasmanova** [comment]
@majetideepak Deepak, thank you for reviewing. Updated.

**Review by majetideepak — APPROVED** [review]
Thanks @mbasmanova

**Review by aditi-pandit — APPROVED** [review]

**Comment by aditi-pandit** [comment]
Thanks Masha for putting this together. The explanation really helps.

File: velox/docs/develop/task.rst [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+nodes and operators. Each plan node with multiple sources introduces a new
+pipeline or pipelines. These plan nodes include HashJoinNode, CrossJoinNode,
+MergeJoinNode, LocalMergeNode, LocalPartitionNode. Join nodes have two sources
+and introduce one new pipeline that ends at the right-side source.
+LocalMergeNode and LocalPartitionNode introduce one pipeline for each source.
+LocalPartitionNode is an exception as it may have a single source in which case
+it still introduces a new pipeline to allow for different parallelism before
+and after. Each pipeline may run multi-threaded and the number of threads used
+for each pipeline can be different. Driver refers to a single thread of
+execution of a single pipeline.
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/develop/task.rst:30]: The number of threads is specified in Task::start

```
  /// Starts executing the plan fragment specified in the constructor. If leaf
  /// nodes require splits (e.g. TableScan, Exchange, etc.), these splits can be
  /// added before or after calling start().
  ///
  /// @param maxDrivers Maximum number of drivers / threads used to run a single
  /// pipeline. Some pipelines are running in fewer threads if they contain plan
  /// nodes that do not support parallelism (e.g. final limit) or their source
  /// nodes require splits and there are not enough of these.
  /// @param concurrentSplitGroups In grouped execution, maximum number of
  /// splits groups processed concurrently.
  static void start(
      std::shared_ptr<Task> self,
      uint32_t maxDrivers,
      uint32_t concurrentSplitGroups = 1);
```

File: velox/docs/develop/task.rst [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    :width: 400
+    :align: center
+
+LocalPartitionNode is used to (1) repartition the data, or (2) gather data from
+multiple threads into a single thread, or (3) split the data produced by a
+single thread to allow for multi-threaded processing. A query plan with a
+LocalPartitionNode is converted into two pipelines: a producer pipeline that
+ends with LocalPartition operator and a consumer pipeline that starts with a
+LocalExchangeSourceOperator. These pipelines may run with same or different
+parallelism allowing for N:1, 1:N and N:M local exchange.
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/develop/task.rst:50]: Yes. LocalPartitionNode represents local exchange.

File: velox/docs/develop/task.rst [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+multiple TableScan operators, one per thread of execution (Driver), while
+HashJoinNode is converted into multiple HashProbe operators and multiple
+HashBuild operators. The number of HashProbe and HashBuild operators may or may
+not match.
+
+Task creates Drivers and Operators, then transfers the ownership of Operators to
+the corresponding Drivers. Each Driver is assigned a pipeline ID and a driver
+ID. Pipeline IDs are zero-based sequential numbers. Driver IDs are also
+zero-based sequential numbers scoped to a pipeline. Driver IDs are unique
+within a pipeline, but repeat across pipelines. Each Operator is provided with
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/develop/task.rst:81]: Driver's unique ID can be composed as "pipelineId.driverId". Does that work for the use case you have in mind?

**Comment by mbasmanova** [comment]
@aditi-pandit Aditi, thank you for your feedback. Updated.

**Review by aditi-pandit — APPROVED** [review]

**Review by aditi-pandit — APPROVED** [review]
Thanks Masha ! Appreciate it.

**Pull request closed without merging**